tests/test_s3utils.py

```diff
@@ -43,50 +42,41 @@ def mock_s3(cls):
 class S3UtilsTestCase(unittest.TestCase):
     """Test for the S3 related utilities.
     """
-    bucketName = "testBucketName"
+    bucketName = "test_bucket_name"
     fileName = "testFileName"
 
     def setUp(self):
         # set up some fake credentials if they do not exist
         self.usingDummyCredentials = setAwsEnvCredentials()
 
-        s3 = boto3.client("s3")
+        self.client = getClient()
         try:
-            s3.create_bucket(Bucket=self.bucketName)
-            s3.put_object(Bucket=self.bucketName, Key=self.fileName,
-                          Body=b"test content")
-        except s3.exceptions.BucketAlreadyExists:
+            self.client.create_bucket(Bucket=self.bucketName)
+            self.client.put_object(Bucket=self.bucketName, Key=self.fileName,
+                                   Body=b"test content")
+        except self.client.exceptions.BucketAlreadyExists:
             pass
 
     def tearDown(self):
-        s3 = boto3.resource('s3')
-        bucket = s3.Bucket(self.bucketName)
-        try:
-            bucket.objects.all().delete()
-        except botocore.exceptions.ClientError as err:
-            errorcode = err.response["ResponseMetadata"]["HTTPStatusCode"]
-            if errorcode == 404:
-                # the key does not exists - pass
-                pass
-            else:
-                raise
-
-        bucket = s3.Bucket(self.bucketName)
-        bucket.delete()
+        objects = self.client.list_objects(Bucket=self.bucketName)
```

Comment:
This only fetches the first 1000 objects. I assume that in tests this should not be a problem, since I assume it's probably safe to assume that number of objects put into the buckets in tests will always be much less than 1000, but still wanted to point it out just in case.

Reply:
Thanks for pointing it out. I saw that limit but thought it's okay for the unit test.  I'll add a comment here to be more clear. Strictly speaking the tests don't need changes because it's done through moto.  Nonetheless the new approach should work on both AWS & GCS assuming there are not that many objects.

Reply:
From the boto3 document it looks like the approach on master via the object collection (`bucket.objects.all().delete()`) has the limit of 1000 too? 
It seems either is adequate for the unit tests here for cleaning up.

Reply:
Hmm, maybe I didn't understand the documentation myself then. This is what motivated that particular use: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Bucket.objects

 > **boto3.Bucket.objects**
> A collection of ObjectSummary resources.A ObjectSummary Collection will include all resources by default, and extreme caution should be taken when performing actions on all resources.

Reply:
I don't require you to change the code, I just wanted to point out that the purpose of SDKs is to wrap the base API calls in a way in which they would behave in a more pleasant and expected way. So paging, caching, not losing overhead on opening a new connection on every request etc. 

On one side an unknown percentage of this could be lost for GCE when current workaround is applied. On the other side I would urge you to be maximally careful not to "un-generalize" the S3datastore by changing calls without thinking through how these things behave in the background is all. That was the point of yesterdays objections. 

Reading back to yesterdays posts I can see how my objections might seem very opinionated, but I don't see anything in this PR that breaks anything else. I certainly also didn't intend the objections in that strong of a way, just as a caution sign. It's fine for me if this goes forward as is. It seems to work for everyone else too.

 [Here](https://github.com/boto/boto/blob/develop/boto/s3/bucket.py#L245) and [Here](https://github.com/boto/boto/blob/91ba037e54ef521c379263b0ac769c66182527d7/boto/s3/bucketlistresultset.py#L43) you can see how paging is implemented, so it does return all objects (think in the lines of `aws s3 --recursive` type of command).